phpBB/phpbb/template/twig/extension/icon.php

```diff
@@ -0,0 +1,321 @@
+<?php
+/**
+ *
+ * This file is part of the phpBB Forum Software package.
+ *
+ * @copyright (c) phpBB Limited <https://www.phpbb.com>
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\template\twig\extension;
+
+use phpbb\template\twig\environment;
+
+class icon extends \Twig\Extension\AbstractExtension
+{
+	/** @var \phpbb\user */
+	protected $user;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param \phpbb\user	$user			User object
+	 */
+	public function __construct(\phpbb\user $user)
+	{
+		$this->user = $user;
+	}
+
+	/**
+	 * Returns the name of this extension.
+	 *
+	 * @return string						The extension name
+	 */
+	public function getName()
+	{
+		return 'icon';
+	}
+
+	/**
+	 * Returns a list of functions to add to the existing list.
+	 *
+	 * @return \Twig\TwigFunction[]			Array of twig functions
+	 */
+	public function getFunctions()
+	{
+		return [
+			new \Twig\TwigFunction('Icon', [$this, 'icon'], ['needs_environment' => true]),
+		];
+	}
+
+	/**
+	 * Generate icon HTML for use in the template, depending on the mode.
+	 *
+	 * @param environment	$environment	Twig environment object
+	 * @param string		$type			Icon type (font|iconify|png|svg)
+	 * @param string		$icon			Icon name (eg. "bold")
+	 * @param string		$title			Icon title
+	 * @param bool			$hidden			Hide the icon title from view
+	 * @param string		$classes		Additional classes (eg. "fa-fw")
+	 * @param array			$attributes		Additional attributes for the icon, where the key is the attribute.
+	 *                      				{'data-ajax': 'mark_forums'} results in ' data-ajax="mark_forums"'
+	 * @return string
+	 */
+	public function icon(environment $environment, $type, $icon, $title = '', $hidden = false, $classes = '', array $attributes = [])
```

Comment:
So, currently you need to explicitly define icon type to render it. Can we add some `auto` mode to automatically detect icon type and render it then? Or even split such code into a function and then use it here instead of `switch`.

Reply:
No that makes no sense and defeats the purpose.  How would you even automatically detect an icon type? There are too different.

Reply:
Automatically detecting will also not return one definitive result, as a lot of the icon sets use the same names. So, that would only be possible indeed if some 'default' was configured. And then you can just use that variable in the template instead: `{{ icon(DEFAULT_ICON, 'user') }}`.